sulong/projects/com.oracle.truffle.llvm.nodes/src/com/oracle/truffle/llvm/nodes/intrinsics/multithreading/UtilStartThread.java

```diff
@@ -0,0 +1,110 @@
+import com.oracle.truffle.api.Truffle;
+import com.oracle.truffle.api.TruffleLanguage;
+import com.oracle.truffle.api.frame.FrameSlot;
+import com.oracle.truffle.api.frame.VirtualFrame;
+import com.oracle.truffle.api.nodes.RootNode;
+import com.oracle.truffle.llvm.runtime.LLVMContext;
+import com.oracle.truffle.llvm.runtime.LLVMLanguage;
+import com.oracle.truffle.llvm.runtime.memory.LLVMStack;
+import com.oracle.truffle.llvm.runtime.nodes.api.LLVMExpressionNode;
+import com.oracle.truffle.llvm.runtime.types.FunctionType;
+import com.oracle.truffle.llvm.runtime.types.PointerType;
+import com.oracle.truffle.llvm.runtime.types.Type;
+
+public class UtilStartThread {
+    static class InitStartOfNewThread implements Runnable {
+        private Object startRoutine;
+        private Object arg;
+        private TruffleLanguage.ContextReference<LLVMContext> ctxRef;
+
+        public InitStartOfNewThread(Object startRoutine, Object arg, TruffleLanguage.ContextReference<LLVMContext> ctxRef) {
+            this.startRoutine = startRoutine;
+            this.arg = arg;
+            this.ctxRef = ctxRef;
+        }
+
+        @Override
+        public void run() {
+            if (ctxRef.get().pthreadCallTarget == null) {
+                ctxRef.get().pthreadCallTarget = Truffle.getRuntime().createCallTarget(new RunNewThreadNode(LLVMLanguage.getLanguage()));
+            }
+            // pthread_exit throws a control flow exception to stop the thread
+            try {
+                Object retVal = ctxRef.get().pthreadCallTarget.call(startRoutine, arg);
+                // no null values in concurrent hash map allowed
+                if (retVal == null) {
+                    retVal = 0;
+                }
+                ctxRef.get().retValStorage.put(Thread.currentThread().getId(), retVal);
+            } catch (PThreadExitException e) {
+
+            }
+        }
+    }
+
+    static final class MyArgNode extends LLVMExpressionNode {
+        private final FrameSlot slot;
+
+        private MyArgNode(FrameSlot slot) {
+            this.slot = slot;
+        }
+
+        @Override
+        public Object executeGeneric(VirtualFrame frame) {
+            return frame.getValue(slot);
+        }
+    }
+
+    private static class RunNewThreadNode extends RootNode {
+        @Child
+        LLVMExpressionNode callNode = null;
+
+        @CompilerDirectives.CompilationFinal
+        FrameSlot functionSlot = null;
+
+        @CompilerDirectives.CompilationFinal
+        FrameSlot argSlot = null;
+
+        @CompilerDirectives.CompilationFinal
+        FrameSlot spSlot = null;
+
+        private LLVMLanguage language;
+
+        protected RunNewThreadNode(LLVMLanguage language) {
+            super(language);
+            this.language = language;
+        }
+
+        @Override
+        public Object execute(VirtualFrame frame) {
+            LLVMStack.StackPointer sp = language.getContextReference().get().getThreadingStack().getStack().newFrame();
```

Comment:
Better use a stored `ContextReference` instead of getting it at every execution.

Reply:
ok, i'm using a stored context now. i hope that's also fine, as you suggested earlier it's mostly more efficient to use context directly instead of using a context reference.